### ISSUE TITLE
Add a 'fix' command to automatically update dependencies.

### ISSUE THREAD

**Issue by phansch** [RustSec/cargo-audit#23]
Similar to https://docs.npmjs.com/cli/audit

Running `cargo audit fix` or a similar command would install all compatible (as in SemVer compatible, no automatic major version bumps) updates to all vulnerable packages.
This would also rewrite version constraints in `Cargo.toml`.
I realize this may be a bigger issue, I just wanted to get this idea out there.

**Comment by tarcieri** [comment]
It's a good idea. I'll leave the issue open in case anyone wants to contribute it.

**Comment by zofrex** [comment]
Any thoughts on how this could best be implemented?

Should it call out `cargo` to update the deps, or edit the .toml/.lock files?

**Comment by tarcieri** [comment]
Any prior art to look at here?

**Comment by freddyb** [comment]
[`cargo update`](https://doc.rust-lang.org/cargo/commands/cargo-update.html) seems to cover this, no?

**Comment by tarcieri** [comment]
@freddyb only in cases where dependencies aren't locked to all vulnerable versions in Cargo.toml

**Comment by hbina** [comment]
What is the status of this issue? I would like to pick it up

**Comment by tarcieri** [comment]
@hbina we're looking for interested contributors! There's no concrete plans for anyone to work on this yet.

I can provide a bit of guidance for getting started. You'll want to leverage the dependency tree features in the `cargo-lock` crate. Here's an example of where `cargo-audit` is presently using them (to print the inverse dependency trees):

https://github.com/RustSec/cargo-audit/blob/master/src/presenter.rs#L65

Here's the documentation on `cargo_lock::dependency::Tree`:

https://docs.rs/cargo-lock/latest/cargo_lock/dependency/tree/struct.Tree.html

This uses the [`petgraph`](https://docs.rs/petgraph/) crate to model the dependency tree as a directed acyclic graph. Here's an example of the code that walks the dependency tree in order to print the graph:

https://github.com/RustSec/cargo-lock/blob/master/src/dependency/tree.rs#L161

For each vulnerable dependency, you'll want to do an inverse walk of the dependency tree (`petgraph::Direction::Incoming`) until you reach the "edge" packages. These theoretically should be ones defined in the local workspace. You'll want to make note of the packages immediately before the "edge" ones which are depended upon by them: these are the ones that need to be upgraded.

For example, given this tree:

```
vulnerable-package 1.2.3
└── bar 0.1.2
    └── foo 0.0.1
```

`foo` should be a package in the local workspace, which depends on `bar` that needs upgrading.

From there we'll need a (workspace aware) `Cargo.toml` parser, which is able to locate the relevant `Cargo.lock` for the `foo` package in the workspace, and then examine the dependencies section to determine what the requirements are. It will probably also need to hit the crates.io API (which we're presently doing with the [`crates_io_api`](https://github.com/theduke/crates_io_api) crate) in order to determine a suitable upgrade path.

Let me know if that makes sense, and if you have any questions about how to proceed.

**Comment by gibfahn** [comment]
>Any prior art to look at here? (i.e. Rust tools which bump dependencies directly in Cargo.toml)

[`cargo upgrade`](https://github.com/killercup/cargo-edit#cargo-upgrade) does Cargo.toml updates.

**Comment by tarcieri** [comment]
Indeed! The `cargo-edit` crate has the sort of functionality we're after, and it looks like it may even be possible to consume it as a library:

https://docs.rs/cargo-edit/

**Comment by hbina** [comment]
I tried including it as a dependency but there's a conflict in `Cargo.toml` regarding `libgit2-sys`.
```
hbina085@akarin:~/git/cargo-audit$ cargo build
    Updating crates.io index
error: failed to select a version for `libgit2-sys`.
    ... required by package `git2 v0.10.0`
    ... which is depended on by `rustsec v0.15.0`
    ... which is depended on by `cargo-audit v0.9.2 (/home/hbina085/git/cargo-audit)`
versions that meet the requirements `^0.9.0` are: 0.9.1, 0.9.0

the package `libgit2-sys` links to the native library `git2`, but it conflicts with a previous package which links to `git2` as well:
package `libgit2-sys v0.8.2`
    ... which is depended on by `git2 v0.9.2`
    ... which is depended on by `cargo-edit v0.4.0`
    ... which is depended on by `cargo-audit v0.9.2 (/home/hbina085/git/cargo-audit)`

failed to select a version for `libgit2-sys` which could resolve this conflict
hbina085@akarin:~/git/cargo-audit$ cargo build
    Updating crates.io index
error: failed to select a version for `libgit2-sys`.
    ... required by package `git2 v0.10.0`
    ... which is depended on by `rustsec v0.15.0`
    ... which is depended on by `cargo-audit v0.9.2 (/home/hbina085/git/cargo-audit)`
versions that meet the requirements `^0.9.0` are: 0.9.1, 0.9.0

the package `libgit2-sys` links to the native library `git2`, but it conflicts with a previous package which links to `git2` as well:
package `libgit2-sys v0.8.2`
    ... which is depended on by `git2 v0.9.2`
    ... which is depended on by `cargo-edit v0.4.0`
    ... which is depended on by `cargo-audit v0.9.2 (/home/hbina085/git/cargo-audit)`

failed to select a version for `libgit2-sys` which could resolve this conflict
```

**Comment by tarcieri** [comment]
@hbina looks like they need to bump this to `0.10`:

https://github.com/killercup/cargo-edit/blob/master/Cargo.toml#L67

**Comment by tarcieri** [comment]
I opened a PR to bump the `git2` version used by `cargo-edit`: https://github.com/killercup/cargo-edit/pull/341

**Comment by tarcieri** [comment]
@hbina the PR to upgrade `cargo-edit` has been merged, if you want to temporarily include it via git to experiment with

**Comment by hbina** [comment]
@tarcieri Is there a convenient way to convert a `VersionReq` to `String`s? AFAIK, even `cargo-edit` uses raw string/`String` to deal with dependency stuff (`semver` is only used for its ordering capability, not as an argument for a method, for example)

**Comment by tarcieri** [comment]
Where did you open the PR? I don't see it.

**Comment by hbina** [comment]
My bad, I was still editing the PR message.

**Comment by aug2uag** [comment]
hi there i'm just doing some practice with Rust, and also trying to familiarize myself more with security-- i've spent a little time on this issue for these purposes and have a flow proposal i'd like to share for your evaluation and feedback

before i finish, i was wondering if i could get some feedback how to pass the Vulnerability data, namely package name for `cargo-edit`

the business end of my implementation has the Vulnerability in the function signature, and i was thinking to implement something similar to `generate_lockfile` with `Command::new` to the `edit-cargo` package

```
    /// try `cargo-edit` upgrade
    pub fn upgrade_vulnerability(&self, arg: &Vulnerability) {
            println!("{:?}", arg);
            unimplemented!();
    }
```

**Comment by tarcieri** [comment]
@aug2uag note that there's a PR for this feature already open:

https://github.com/RustSec/cargo-audit/pull/151

It takes a slightly different approach from yours and pulls in [`cargo-edit`] as a library, rather than calling out to it as a subcommand.

These are debatable tradeoffs: using it as a library provides a more seamless experience, however it also adds to build times, and some people may not want the `fix` feature. Another alternative there is to make `fix` an optional Cargo feature, so people who don't want it and would prefer faster builds can disable it.

**Comment by aug2uag** [comment]
Thank you @tarcieri , indeed @hbina very nice and elegant implementation. Lots of good data there.

It wasn't my intention to have the upgrade occur as a library, absolutely from the comment that's the case, and my intention was to have it used as a subcommand. Your observation Tony is interesting, great analysis.

My implementation had `fix` as a subcommand added in addition to `audit` in `CargoAuditCommand` with an update to the CLI that causes enumeration on args for match with "fix" to initiate the audit sequence with a conditional on where vulnerabilities are evaluated in `presenter`. The `OutputConfig`  struct was changed to handle an additional boolean to define fix and upgrade state. 

Thanks for the info, and I hope I can be of assistance to you on other issues.

**Comment by tarcieri** [comment]
@aug2uag thanks for taking a look! And I'd be interested in comparing these two approaches and getting people's thoughts

**Comment by tarcieri** [comment]
@hbina @aug2uag `cargo-edit` 0.4.1 is now out and uses the same version of `git2` as `cargo-audit`:

https://github.com/killercup/cargo-edit/releases/tag/v0.4.1

**Comment by tarcieri** [comment]
@hbina @aug2uag now that `cargo-edit` is synced up with our dependency versions, I would like to pick one of your two PRs to use as the basis of a `fix` feature.

Upon cursory review, @aug2uag's PR (#157) seems slightly more complete in terms of functionality. Curious to hear both of your thoughts on the matter though.

**Comment by tarcieri** [comment]
I just landed #157, which adds initial `fix` support. Closing this issue
